### ISSUE TITLE
Revisit the BlockFetch / Chainsync design to better cater for parallel downloads.

### ISSUE THREAD

**Issue by karknu** [input-output-hk/ouroboros-network#1526]
The current design makes handling of blocks out of order rather costly, #1393 & #1525.

**Comment by edsko** [comment]
Note: blocks arriving out of order will result in a slow-down in the consensus layer. When blocks arrive _in_ order, each block is considered in turn, added to the chain, written to disk (not flushed), and that's the end of it. When blocks arrive out of order, however, they must be reconsidered later, which amongst other things involves reading them back from disk. We have a ticket for introducing a small cache here (https://github.com/input-output-hk/ouroboros-network/issues/1383), but at the moment that is not happening. 

It's possible that is a partial explanation for these slowdowns.

**Comment by dcoutts** [comment]
Yes, this ticket is for us to revisit the design here, and either improve things on the chain db side, or block fetch side, whichever we all agree is the better approach. We don't need this in the federated setting, but it'll become more important in the p2p setting, so labelling the milestone accordingly.
